proposals/0000-th-stage-hygiene.rst

```diff
@@ -0,0 +1,397 @@
+  File formats are independent of RTS ABIs, and so the stage 1 compiler can always load a "fat" interface file it itself created and interpret it.
+  Putting everything together, the stage 1 compiler makes native "fat" interface files for "stage 2, TH stage -1", and splices their evaluations into the "stage 2, TH stage 0" code to make the stage 2 GHC that will run on the foreign platform.
+  [N.B. In the easy case when we don't change the ABI, compiler bootstrapping stages and TH stages coincide!
+  Stage 2 - 1 = Stage 1.]
+
+  Switching all existing ``derive-*`` code generators to TH would probably make them lighter and easier to maintain.
+  It should also allow building GHC the binary with plain `cabal`.
+  Hadrian would be one step closer to being another implementation of Cabal/cabal-install without GHC-specific logic.
+
+``*_BUILD_*`` and ``*_HOST_*`` not always defined
+  The conditional definition of the CPP macros ensures they don't pollute the purity of the build when they don't matter.
+  This is important for highly pure build systems like Nix to not have to needless rebuild stuff when the target platform changes.
+  It will also cut down on people improperly using "target" when they meant "host".
+
+``Lift`` and qualified goals
+  When we carefully introduce lifting to stage-hygienic goals, we need to ensure that the type being lifted is the same or "close enough".
+  This means we will need to introduce a intra-package constraint on package defining that type across the stage pair where Lift is made available.
+  Conceptually, there might be an auto-generated package with the orphan ``Lift`` instance which imposes the same version constraint on it's library dependency in both stages.
+
+  In particular, existing qualified dependencies from ``setup-depends`` and ``build-tool-depends`` are from stage *n* to *n - 1*;
+  that the stages are different alone explains why versions are allowed to differ.
+  In particular this means given a dependency edge where the needed and needing components are in the same package regardless of their relative stage indices,
+  the same version of the package must be used for both.
+
+
+Costs and Drawbacks
+-------------------
+
+- This is a huge amount of work.
+  But I am fine chipping away it over a long period of time.
+
+- Even a temporary conflict between typed TH and this could slow typed TH's adoption.
+
+- I don't know of precedent for extensions that prevent modules from being linked together.
+
+- Most existing libraries with commonly used TH helpers (`lens`, `aeson`) have the TH in the same call component but in a different module.
+  To leverage this proposal, we would have to refactor them to put those modules in a separate library component.
+  It would take decent amount of conditional code to still support old GHCs, and even more to not be a breaking change on those old libraries.
+
+Alternatives
```

Comment:
Are there any alternatives that give us the benefit without introducing this infinite set of stages? Or without exposing them as much to the user? I can see stages have significant precedent, but I'd be interested to hear if there are any other approaches.

Reply:
I in fact restrict to stage -1, 0, 1 in the proposed change specification. But (perhaps confusingly) all the design components assumes the full integers to hold them to a higher standard. I don't want anything to be allowed to be sloppy or ad-hoc by doing a 3 stage case analysis, rather than having to deal uniformly with arbitrarily many stages.